### PR TITLE
feat(dynamic sampling): Implements tasks for generating transaction boosting rules [TET-519]

### PR DESCRIPTION
**Pull request by RaduW**
This PR implements the infrastructure that gets the data for generating transaction boosting rules.
It implements the tasks that collect transaction boosting information and store it in redis

### CODE REVIEW

File: src/sentry/dynamic_sampling/rules/helpers/prioritize_transactions.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+from typing import Mapping, Tuple, cast
+
+from sentry.dynamic_sampling.rules.utils import get_redis_client_for_ds
+from sentry.utils import json
+
+
+def _get_cache_key(org_id: int, proj_id: int) -> str:
+    return f"ds::o:{org_id}:p:{proj_id}:pri_tran"
```
Comment (andriisoldatenko) [inline src/sentry/dynamic_sampling/rules/helpers/prioritize_transactions.py:8]: remember we decided to use a short version (without project_id):
https://github.com/getsentry/sentry/blob/41ff6d0ca205a37ce74d49780781136467a5f7d6/src/sentry/dynamic_sampling/rules/helpers/prioritise_project.py#L9-L10

and later use `hset` instead.

File: src/sentry/dynamic_sampling/tasks.py [hunk 2]
```diff
@@ -102,3 +117,90 @@ def adjust_sample_rates(
...
+                    large_transactions=False,
+                    max_transactions=MAX_TRANSACTIONS_PER_PROJECT // 2,
+                ),
+            ):
+
+                if not current_org or current_org.id != project_transactions["org_id"]:
+                    current_org = Organization.objects.get_from_cache(
+                        id=project_transactions["org_id"]
+                    )
+                    current_org_enabled = features.has(
```
Comment (evanh) [inline src/sentry/dynamic_sampling/tasks.py:159]: I would be inclined to move this check before the transactions queries. Why fetch and process all the transactions if we are just going to throw them away?

**Review by evanh — COMMENTED** [review]
This generally looks good to me. Added one small comment.

File: src/sentry/dynamic_sampling/tasks.py (hunk 2 above)
Comment (RaduW) [reply src/sentry/dynamic_sampling/tasks.py:159]: Thanks @evanh  for you input.
Yes, you are absolutely right, we have debated about checking early and getting the transactions only for the
orgs that are enabled.

The reason we decided to go for this is that we wanted to check if we are able to run the queries for everything and for now only apply the changes to a limited group of orgs (that we use for testing).

The feature is only intended to be temporary, once we are happy with the result on our test orgs we will enable this for everybody and we don't want to have the unpleasant surprise that the database cannot handle it.

If we have any doubts about the query performance we can use the same approach that we used for project and do the first query ( the query for orgs) using modulo i.e. `Condition(Function("modulo", [Column("org_id"), 100]), Op.LT, load_rate)`. This way we can ramp up db load independently from enabling the organisations for transaction boosting.

WDYT ?

File: src/sentry/dynamic_sampling/prioritise_transactions.py [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+
+from sentry.sentry_metrics import indexer
+from sentry.snuba.dataset import Dataset, EntityKey
+from sentry.snuba.metrics.naming_layer.mri import TransactionMRI
+from sentry.snuba.referrer import Referrer
+from sentry.utils.snuba import raw_snql_query
+
+logger = logging.getLogger(__name__)
+MAX_SECONDS = 60
+CHUNK_SIZE = 1000
```
Comment (onewland) [inline src/sentry/dynamic_sampling/prioritise_transactions.py:29]: do we want to increase `CHUNK_SIZE` to 10000 (Snuba's limit)? or did we turn down that idea for some reason?

https://www.notion.so/sentry/Project-bias-analysis-44a0c93358c94979b18ebd06a6b7f52e?pvs=4#34a2cff938f04d34a4b62a1c75acaadd

**Review by onewland — COMMENTED** [review]
Mostly seems good to me, but if we're not increasing `CHUNK_SIZE` before merge I'd like to know the reason

File: src/sentry/dynamic_sampling/tasks.py (hunk 2 above)
Comment (RaduW) [reply src/sentry/dynamic_sampling/tasks.py:159]: Done.

File: src/sentry/dynamic_sampling/prioritise_transactions.py (hunk 3 above)
Comment (RaduW) [reply src/sentry/dynamic_sampling/prioritise_transactions.py:29]: Yes thanks, I missed that.

File: src/sentry/dynamic_sampling/prioritise_transactions.py [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+                match=Entity(EntityKey.GenericOrgMetricsCounters.value),
+                select=[
+                    Function("uniq", [Column("project_id")], "num_projects"),
+                    Column("org_id"),
+                ],
+                groupby=[
+                    Column("org_id"),
+                ],
+                where=[
+                    Condition(Function("modulo", [Column("org_id"), 100]), Op.LT, load_rate),
```
Comment (nikhars) [inline src/sentry/dynamic_sampling/prioritise_transactions.py:64]: This is fine for now. When you decide to change this to 100% and are satisfied with the results, please remove this modulo clause altogether.

**Review by nikhars — APPROVED** [review]
🔥

**Review by onewland — APPROVED** [review]

File: src/sentry/dynamic_sampling/prioritise_transactions.py (hunk 4 above)
Comment (RaduW) [reply src/sentry/dynamic_sampling/prioritise_transactions.py:64]: Yes, thanks @nikhars, this is the plan.

File: src/sentry/dynamic_sampling/prioritise_transactions.py [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+    offset = 0
+    org_ids = list(org_ids)  # just to be sure it is not some other sequence
+    transaction_string_id = indexer.resolve_shared_org("transaction")
+    transaction_tag = f"tags_raw[{transaction_string_id}]"
+    metric_id = indexer.resolve_shared_org(str(TransactionMRI.COUNT_PER_ROOT_PROJECT.value))
+    current_org_id: Optional[int] = None
+    current_proj_id: Optional[int] = None
+    transaction_counts: List[Tuple[str, int]] = []
+
+    if large_transactions:
```
Comment (andriisoldatenko) [inline src/sentry/dynamic_sampling/prioritise_transactions.py:133]:
```suggestion
    if large_transactions_first:
```

File: src/sentry/dynamic_sampling/rules/helpers/prioritize_transactions.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    redis_client = get_redis_client_for_ds()
+    cache_key = _get_cache_key(org_id=org_id, proj_id=proj_id)
+    try:
+        serialised_val = redis_client.get(cache_key)
+        if serialised_val:
+            val = json.loads(serialised_val)
+            ret_val = cast(Tuple[Mapping[str, float], float], val)
+            return ret_val
+    except (TypeError, ValueError):
+        # for any error just get the default
```
Comment (andriisoldatenko) [inline src/sentry/dynamic_sampling/rules/helpers/prioritize_transactions.py:23]: do we what to capture exceptions here by sentry?

File: src/sentry/dynamic_sampling/prioritise_transactions.py (hunk 5 above)
Comment (RaduW) [reply src/sentry/dynamic_sampling/prioritise_transactions.py:133]: The fact that the large transactions come first is not the point, the point of the parameter is that the function fetches large_transactions ( i.e. the sorting is not relevant, what the function returns is).

**Review by andriisoldatenko — APPROVED** [review]

File: src/sentry/dynamic_sampling/rules/helpers/prioritize_transactions.py (hunk 6 above)
Comment (RaduW) [reply src/sentry/dynamic_sampling/rules/helpers/prioritize_transactions.py:23]: Don't remember exactly why I have both, but it only catches deserialisation errors.

Comment (andriisoldatenko) [reply src/sentry/dynamic_sampling/rules/helpers/prioritize_transactions.py:23]: `sentry_sdk.capture_exception(e)`

**Pull request merged**